## Power Method (Normalised) 

In [14]:
from pylab import *
import numpy as np

A=np.array([[19,-1,-2,2,0.5],[-1,5,-1,-2,1],[-2,-1,2,-1,5],[2,-2,-1,3,0.1],[50,-1,-2,2,2]],dtype=np.float64)
print('The matrix A = \n',A ,'\n')
X=np.array([-1,-1,0,0,0],dtype=np.float64)
b=np.argmax(np.abs(X))
c=X[b]
Y=X/X[b] 

m=np.linalg.eigvals(A)
print('The Eigenvalues of A by LinAlg are: \n' , m , '\n')

e , n = 1.0 , 0
while e>0.0001:
    X=np.dot(A,Y)
    b1=np.argmax(np.abs(X))
    c1=X[b1]
    Y=X/X[b1]
    e=abs(c1-c)
    c=c1
    print(f'The eigenvalue in step {n+1} is: ',c, '\n')
    n+=1
    
print(f'The dominant eigenvalue is {c}')
print(f'The corresponding eigenvector is {Y}')
print(f'Number of steps taken - {n}')

The matrix A = 
 [[19.  -1.  -2.   2.   0.5]
 [-1.   5.  -1.  -2.   1. ]
 [-2.  -1.   2.  -1.   5. ]
 [ 2.  -2.  -1.   3.   0.1]
 [50.  -1.  -2.   2.   2. ]] 

The Eigenvalues of A by LinAlg are: 
 [19.12428569 -2.41663392  1.03897261  7.30349304  5.9498826 ] 

The eigenvalue in step 1 is:  49.0 

The eigenvalue in step 2 is:  20.408163265306126 

The eigenvalue in step 3 is:  20.0448 

The eigenvalue in step 4 is:  19.339534442848016 

The eigenvalue in step 5 is:  19.20127003993994 

The eigenvalue in step 6 is:  19.14815329038658 

The eigenvalue in step 7 is:  19.132223476511506 

The eigenvalue in step 8 is:  19.12688811870647 

The eigenvalue in step 9 is:  19.125154595872335 

The eigenvalue in step 10 is:  19.124577633351876 

The eigenvalue in step 11 is:  19.124384904952308 

The eigenvalue in step 12 is:  19.12431974208691 

The dominant eigenvalue is 19.12431974208691
The corresponding eigenvector is [0.35161191 0.02391151 0.24768121 0.03148794 1.        ]
Number of steps t

## Power Method (Inverse) 

In [24]:

print('The matrix A = \n',A ,'\n')
a = np.linalg.eigvals(A)
print('The eigenvalues of matrix A by LinAlg are: \n',a,'\n')
B = np.linalg.inv(A)

X=np.array([-1,-1,0,0,0],dtype=np.float64)
b=np.argmax(np.abs(X))
c=X[b]
Y=X/X[b] 

e,n = 1.0,0
while e>0.0001:
    X=np.dot(B,Y)
    b1=np.argmax(np.abs(X))
    c1=X[b1]
    Y=X/X[b1]
    e=abs(c1-c)
    c=c1
    n+=1
    
print(f'The smallest absolute eigenvalue of A is: {1/c} \n')
print(f'The corresponding eigenvector is: {Y} \n')
print(f'Number of steps taken = {n}')

The matrix A = 
 [[19.  -1.  -2.   2.   0.5]
 [-1.   5.  -1.  -2.   1. ]
 [-2.  -1.   2.  -1.   5. ]
 [ 2.  -2.  -1.   3.   0.1]
 [50.  -1.  -2.   2.   2. ]] 

The eigenvalues of matrix A by LinAlg are: 
 [19.12428569 -2.41663392  1.03897261  7.30349304  5.9498826 ] 

The smallest absolute eigenvalue of A is: 1.0389943797611874 

The corresponding eigenvector is: [-0.00273821  0.63440048  0.7057435   1.          0.1901526 ] 

Number of steps taken = 14


## Shifting Power Method 

In [26]:
from pylab import *
import numpy as np

A=np.array([[19,-1,-2,2,0.5],[-1,5,-1,-2,1],[-2,-1,2,-1,5],[2,-2,-1,3,0.1],[50,-1,-2,2,2]],dtype=np.float64)
X=np.array([-1,-1,0,0,0],dtype=np.float64)
b=np.argmax(np.abs(X))
c=X[b]
Y=X/c

m=np.linalg.eigvals(A)
print('The eigenvalues of A by LinAlg are: \n',m,'\n')

e,n = 1.0 , 0
while e>0.0001:
    X=np.dot(A,Y)
    b1=np.argmax(np.abs(X))
    c1=X[b1]
    Y=X/X[b1]
    e=abs(c1-c)
    c=c1
    n+=1
    
print(f'The dominant eigenvalue is: {c} \n')
print(f'The corresponding eigenvector is: {Y} \n')
print(f'Number of steps taken - {n} \n')

I=np.identity(len(A[:,0]),dtype=np.float64)
B= A - c*I

X=np.array([-1,-1,0,0,0],dtype=np.float64)
b1=np.argmax(np.abs(X))
c2=X[b1]
Y=X/X[b1] 

a=np.linalg.eigvals(B)
print('The eigenvalues of B by LinAlg are: \n',a,'\n')

e,n = 1.0,0
while e>0.0001:
    X=np.dot(B,Y)
    b2=np.argmax(np.abs(X))
    c3=X[b2]
    Y=X/X[b2]
    e=abs(c3-c2)
    c2=c3
    n+=1
    
print(f'The smallest eigenvalue of A is: {c2+ c} \n')
print(f'The eigenvector corresponding to this eigenvalue is {Y} \n')
print(f'Number of steps taken - {n}')


The eigenvalues of A by LinAlg are: 
 [19.12428569 -2.41663392  1.03897261  7.30349304  5.9498826 ] 

The dominant eigenvalue is: 19.12431974208691 

The corresponding eigenvector is: [0.35161191 0.02391151 0.24768121 0.03148794 1.        ] 

Number of steps taken - 12 

The eigenvalues of B by LinAlg are: 
 [-3.40564263e-05 -2.15409537e+01 -1.80853471e+01 -1.18208267e+01
 -1.31744371e+01] 

The smallest eigenvalue of A is: -2.4171504566854374 

The eigenvector corresponding to this eigenvalue is [ 0.09914097  0.32116838  1.          0.27984094 -0.72355527] 

Number of steps taken - 43


## Use of Linear Algebra for finding last two eigenvalues

In [97]:
from pylab import *
import numpy as np

A = np.array([[19,-1,-2,2,0.5],[-1,5,-1,-2,1],[-2,-1,2,-1,5],[2,-2,-1,3,0.1],[50,-1,-2,2,2]],dtype=np.float64)
N = 5
## From Linear Algebra we know that sum of all the diagonal entries of a diagonalizable matrix = sum of all of its eigen values
sum = 0.0
for i in range(N):
    sum = sum + A[i][i]
# We have already found 3 eigenvalues so, let m and n be the 2 unknown eigenvalues:
# m + n = sum -  19.12428569 - (−2.41663392) - 1.03897261 . Therefore m+n = 13.25
# Also from linear algebra we know that product of all of its eigenvalues = Det(A)
Det  = np.linalg.det(A)
print('sum = ',sum,'and','Determinant = ',Det)
# So, Det = 19.12428569 x (−2.41663392) x 1.03897261 x m x n
# So mxn = 43.45 and m+n = 13.25 . Solving these 2 eq we get x = 7.30 and y = 5.949
print('The last Two eigenvalues are: ', 7.30 ,'and', 5.949)
    


sum =  31.0 and Determinant =  -2086.599999999999
The last Two eigenvalues are:  7.3 and 5.949


## Conclusions
Ans)a) <br>The 5 eigenvalues of A are: <br> $$[19.12428569 , -2.41663392 , 1.03897261 ,  7.30349304 ,  5.9498826 ]$$ <br>
Ans)b) <br>The Dominant Eigenvalue of A is: $19.12428569$ and the corresponding eigenvector is:     
$$[0.35161191 , 0.02391151 , 0.24768121  ,0.03148794 , 1.00000000]$$ <br>
The smallest absolute eigenvalue of A is: $1.0389943797611874$ and the corresponding eigenvector is:
$$[-0.00273821,0.63440048,0.7057435,1.00000000,0.1901526 ] $$ <br>
The smallest eigenvalue of A is: $-2.4171504566854374$ and the corresponding eigenvector is:
$$[ 0.09914097,0.32116838,1.00000000,0.27984094,-0.72355527]$$ <br>
Ans)c) <br>The eigenvalues at $5^{th}$ and $7^{th}$ iteration are: $19.20127003993994$ and $19.132223476511506$ respectively<br>

## Limitations
For any general nxn matrix it won't be possible to determine all the eigenvalues using power method as it is highly dependent on the ratios of the consecutive eigenvalues $\frac{|\lambda_2|}{|\lambda_1|} \approx 1$ so, its not necessary that the method might converge , also repeated eigenvalues can not be computed and in case of zero eigenvalue the method will computationally give a division by zero error. Power method can never computationally give us the -ve( 2nd largest $|\lambda|$) ; Also in Linear Algebra technique, it can not be used for a non-diagonalizable matrix 
